src/dune_engine/build_system.ml

```diff
@@ -30,20 +30,100 @@ end = struct
+
+module Progress = struct
+  type t =
+    { number_of_rules_discovered : int
+    ; number_of_rules_executed : int
+    }
+
+  let equal { number_of_rules_discovered; number_of_rules_executed } t =
+    Int.equal number_of_rules_discovered t.number_of_rules_discovered
+    && Int.equal number_of_rules_executed t.number_of_rules_executed
+
+  let complete t = t.number_of_rules_executed
+
+  let remaining t = t.number_of_rules_discovered - t.number_of_rules_executed
+
+  let is_determined { number_of_rules_discovered; number_of_rules_executed } =
+    number_of_rules_discovered <> 0 || number_of_rules_executed <> 0
+end
 
 module State = struct
+  module Svar = Fiber.Svar
+
+  type t =
+    | Initializing
+    | Building of Progress.t
+    | Interrupting_current_build
+    | Build_finished_with_success_and_waiting
+    | Build_finished_with_failure_and_waiting
+
+  let equal x y =
+    match (x, y) with
+    | Building x, Building y -> Progress.equal x y
+    | Initializing, Initializing
+    | Interrupting_current_build, Interrupting_current_build
+    | ( Build_finished_with_success_and_waiting
+      , Build_finished_with_success_and_waiting )
+    | ( Build_finished_with_failure_and_waiting
+      , Build_finished_with_failure_and_waiting ) ->
+      true
+    | _, _ -> false
+
+  let t = Fiber.Svar.create Initializing
+
   (* This mutable table is safe: it maps paths to lazily created mutexes. *)
   let locks : (Path.t, Fiber.Mutex.t) Table.t = Table.create (module Path) 32
 
   (* This mutex ensures that at most one [run] is running in parallel. *)
   let build_mutex = Fiber.Mutex.create ()
 
-  let rule_done = ref 0
+  let progress_init =
+    { Progress.number_of_rules_discovered = 0; number_of_rules_executed = 0 }
+
+  let reset_progress () = Svar.write t (Building progress_init)
+
+  let set what = Svar.write t what
+
+  let incr_rule_done () =
+    let current = Svar.read t in
+    match current with
+    | Building current ->
+      Svar.write t
+        (Building
+           { current with
+             number_of_rules_executed = current.number_of_rules_executed + 1
+           })
+    | _ -> assert false
+
+  let start_rule () =
+    let current = Svar.read t in
+    match current with
+    | Building current ->
+      Svar.write t
+        (Building
+           { current with
+             number_of_rules_discovered = current.number_of_rules_discovered + 1
+           })
+    | _ -> assert false
+
+  let errors_init = Build_config.Error.Set.empty
 
-  let rule_total = ref 0
+  let errors = Svar.create errors_init
 
-  let errors = ref ([] : Error.t list)
+  let reset_errors () = Svar.write errors { errors_init with stamp = 0 }
+
+  let add_error error =
+    let set =
+      let set = Svar.read errors in
+      let current = Error.Id.Map.set set.current (Error.id error) error in
+      { Error.Set.current
+      ; stamp = set.stamp + 1
+      ; last_event = Some (Add error)
+      }
+    in
+    Svar.write errors set
```

Comment:
I'm a bit nervous about such non-atomic read-then-write state modifications. It feels like we should have an atomic `update` function to make this code more robust (it's probably fine right now but who knows how we refactor things in future).

Reply:
We rely on code outside of fiber binds to be atomic just about everywhere. So I'm not sure there's anything particularly special here.

Reply:
OK, sounds good.